### ISSUE TITLE
watch and compile on save work differently for minified files

### ISSUE THREAD

**Issue by lekoala** [malvahq/vscode-plugin-dartsass#25]
hello again,

since the last updates, I believe there is a change in how min files are handled

compile on save:
- if min file directory is the same, append ".min" suffix
- if it's different, keep original file name (really, i think the most people will have the minified files in the same directory so it's best to consider the first scenario)

when using watchers:
- if it's in the same directory : only one file is generated (no .min file)
- if it's in a different directory, one watcher will open two windows to watch for changes (and compile files without .min suffix)

so it seems that if using the same directory, watch files will not generate the .min files as it should. when using different directories it seems to work fine.

**Comment by akkumar** [comment]
@lekoala - you have captured the scenario and the difference correctly here. 

Just to add - 

> so it seems that if using the same directory, watch files will not generate the .min files as it should. 

This is the behavior I had observed with the "--style compressed" flag passed to the watcher. The output file seems to be the "file.css" whether it is minified or not minified. so this obviously causes a problem with 1 watcher launching 2 processes , minified and not minified , with the same "targetDirectory" or none specified. 

Hence we launch minified files only if targetMinifiedDirectory is different from targetDirectory so as to not step on each other. 


> compile on save:
> if min file directory is the same, append ".min" suffix
> if it's different, keep original file name (really, i think the most people will have the minified files in the same directory so it's best to consider the first scenario)

I am not a fan of this change I had introduced it recently but this was introduced to keep the behavior in sync with the watcher behavior above. 

Of course, for "compile on save" - we have the flexibility to name the output file whatever we want and hence the ".min.css" suffix.  I am not aware of such an option in case of "sass watcher" .

So - really this quirkiness or undesirable name of calling a minified file without the .min.css is really a corollary of trying to be in sync with the names produced by the sass watcher. 

let me know your suggestions regarding the same.

**Comment by akkumar** [comment]
> So - really this quirkiness or undesirable name of calling a minified file without the .min.css is really a corollary of trying to be in sync with the names produced by the sass watcher.

anyway for what it's worth I had logged an issue with [dart-sass #953](https://github.com/sass/dart-sass/issues/953) to see if there is anyway we can specify to have the output files to be having .min.css then we can get away from this inconsistent naming altogether.

**Comment by akkumar** [comment]
> really, i think the most people will have the minified files in the same directory so it's best to consider the first scenario

So - what I have here is:` targetDirectory: out ` and `targetMinifiedDirectory: dist` . In this case, the output by "compile on save" and "sass watcher" ends up being the same without discrepancy.

**Comment by akkumar** [comment]
>  [dart-sass #953 ](https://github.com/sass/dart-sass/issues/953) 

As expected that ticket with dart-sass was closed as being out of scope for that utility / tool . 

@lekoala  - Given the current limitation of sass watcher, feel free to suggest what you would like to see , something that can be also consistent with the "compile on save" functionality too. 

Looking forward to the same.

**Comment by lekoala** [comment]
@akkumar so here is what i think. I also realized that probably, autoprefixer is not running either when using the watcher, which makes another difference in how files are generated. So probably the main issue is to use the watcher itself.

in a tool i was using previously, this is how it was done
https://github.com/oklai/koala/blob/7f0098ee06f4fd174fb60147bba15bdb156ff50f/src/app/scripts/fileWatcher.js
basically, using node watcher feature, scanning files for imports and then basically recompile the file (like the on save feature). advantage : there is only one way to compile file (using js api and autoprefixer or any custom workflow), disadvantage : this require quite a bit of code to handle the watcher (although it may end up being simpler that trying to use that watch feature that is not very flexible anyway).
it would have been great to delegate the feature to a third party extension like this one
https://marketplace.visualstudio.com/items?itemName=appulate.filewatcher
but i don't see how it would properly manage the imports.

i quite like the "min" suffix, is a widely adopted convention (much more than having different directories like /out and /dist) and use by many projects : bootstrap, foundation...
as for the directories themselves, I'd love to have a default settings to compile any file in a scss/sass directory to a css directory at the same level, it's just very messy to mix both types of files in the same directory

**Comment by akkumar** [comment]
> I also realized that probably, autoprefixer is not running either when using the watcher

Yup - this is true. Because the watcher really just delegates to the cmd line and I don't know any easy simple way for autoprefixer to intercept a watch compilation process etc. 

>  I'd love to have a default settings to compile any file in a scss/sass directory to a css directory at the same level, 

So as it stands today , if you save a file, "compile on save" that is, it does trigger saving to .css and .min.css at the same directory as I see it. 

The discrepancy / quirkiness starts when the "watcher" comes into picture. (no autoprefixer as you had mentioned earlier and then no .min.css either ). 

> basically, using node watcher feature, scanning files for imports and then basically recompile the file (like the on save feature). 

Thanks for the link though. I am going to have think about it since redoing the entire watcher functionality again by ourselves becomes very interesting.

**Comment by akkumar** [comment]
> i quite like the "min" suffix, is a widely adopted convention (much more than having different directories like /out and /dist) and use by many projects : bootstrap, foundation...

> I'd love to have a default settings to compile any file in a scss/sass directory to a css directory at the same level, it's just very messy to mix both types of files in the same directory

"both type of files" -> you mean scss and css ?   so for a "compile on save", you prefer to have css (and .min.css ) in a newly created directory at the source folder , say "css" ?

**Comment by akkumar** [comment]
> i quite like the "min" suffix, is a widely adopted convention (much more than having different directories like /out and /dist) and use by many projects : bootstrap, foundation...

I do so too. At present this seems to be a limitation of "sass ---style compressed -watch inputdir:outputdir" as mentioned earlier. 

another option is to ignore the `--style compressed` entirely and use something like [cleancss](https://github.com/jakubpawlowicz/clean-css). 

May be setting some kind of file watcher on the output of scss to css and then running cleancss on that. This way - we don't need to write a full-fledged watcher ourselves but some quirky file watching mechanisms. 

this is just a thought and it is not still easy to get this sorted out given the number of 'file watchers' that would be spawned.

**Comment by lekoala** [comment]
@akkumar i don't know what would be the best way to deal with this. if we can monitor the output of the file watcher, it would be easy to apply extra processing to any recompiled file (since we have a line for each recompilation in a form like Compiled {source} to {dest}.
therefore we could apply autoprefixer and create {dest}.min.css if needed

**Comment by akkumar** [comment]
> if we can monitor the output of the file watcher, 

yup. at this point - this plugin gets very interesting (may be [chokidar](https://github.com/paulmillr/chokidar) ) .  

I am going to have to take some time to think about this and then implement it . not until end of the week at least.

**Comment by akkumar** [comment]
So I had made a newer release v0.5.0 (of the plugin). 

This implements its own watcher (on generated .css files on `targetDirectory` ) using `chokidar` and then runs autoprefixer and minimizer (using [cleancss](https://github.com/jakubpawlowicz/clean-css) library , as mentioned above ) to write the file.min.css sucessfully. 

So far so good. But couple of caveats: 

*) The above mentioned functionality works only when `targetDirectory` is set . If it is not set, then the minification does not happen since the `sass --watch` process that runs in the background for the watcher has some quirks. 

Apparently the `sass --watcher src:src` process "reformats" the generated .min.css file (by `chokidar` by our plugin) and "unminifies" the .min.css file again. This happens if the input and output dir happen to be the same.  Once the output directory is different (and preferably at a level not under `src` ) then this quirk does not seem to happen. 

*) 

>  "I also realized that probably, autoprefixer is not running either when using the watcher"

> Yup - this is true. Because the watcher really just delegates to the cmd line and I don't know any easy simple way for autoprefixer to intercept a watch compilation process etc.

So the above statement still holds true after this release. There is no way for the unminified files to have anything to do with autoprefixer but the minified files do get a pass through autoprefixer though. 

*) I have removed the `targetMinifiedDirectory` configuration property altogether, as it more confusing to the end-user. The generated .css and .min.css live in the same directory - `targetDirectory`.

**Comment by lekoala** [comment]
@akkumar hi! that's some great changes there. I agree there is probably no need for targetMinifiedDirectory if we use the .min.css convention. It also make the whole thing simpler.

what would be great is to have some default convention instead of having to set the targetDirectory. In my case, I often use differents paths for each projects (like /themes/themename/css, /themes/othertheme/css...) and really if I could avoid setting this each time it would be great.

Do you think it could be meaningful to have a setting allow to output any file in ./scss or ./sass to ./css or ./dist as the default behaviour ? it seems very logical.. maybe a setting that could be "outputToParentFolder" where you choose the name (css or dist in this case). Maybe it could be a bit tricky for subfolders (./scss/mylib/somefile.scss) because it would need to output to (./css/mylib/somefile.css) and consider the "scss" folder as the "root" folder, but I think it's manageable. What matters most is that they stay on the same level to avoid messing with paths (to images, fonts, etc) which is probably what most people need.
What do you think?

**Comment by akkumar** [comment]
You have definitely brought some very good suggestions. 

My only worry about making assumptions is that different people have different naming conventions and if we are to assume any outputDirectory name , what if that collides with what they already have ?  

> and really if I could avoid setting this each time it would be great.
> maybe a setting that could be "outputToParentFolder" where you choose the name (css or dist in this case

One suggestion that I can think of is to set `dartsass.targetDirectory` at the user level to be `out or dist`   (This would be assumed by a relative path in every project / workspace so setting a relative path should be ok). 

Being vscode  user property, that gets inherited by every workspace as well , so there will be no need to set for every project .

Would that work for you for now ? 

I am trying to ingest your suggestions and see how we can translate it to something more simpler and intuitive at the same time.

**Comment by lekoala** [comment]
@akkumar  yes I also think a relative directory should be fine. ./css would be perfect for me :-)

**Comment by akkumar** [comment]
> ./css would be perfect for me

as it stands today, the plugin supports relative path ( relative to projectRoot) for `sassBinPath` , `targetDirectory` and `includePath` properties. 

You can just make 'targetDirectory: "css"`  to keep it simple.
